dsl: Expand derivative

JDBetteridge opened the pull request:
I want to be able to manipulate Devito derivative objects, more like sympy derivative objects.

(possibly only optionally)

JDBetteridge commented on devito/finite_differences/differentiable.py:
```diff
@@ -263,7 +263,16 @@ def __neg__(self):
-        ret = super().__eq__(other)
+        from devito.finite_differences import Derivative
+        # Prevent recustion if other is also of derivative type
+        if type(self) is type(other) is Derivative:
+            # `sympy.Derivative.__eq__(self, other)` also recurses
+            ret = self.expr == other.expr
+            ret &= self.dimensions == other.dimensions
+            ret &= self.derivative_count == other.derivative_count
+            ret &= self.deriv_order == other.deriv_order
```
This may or may not be sufficient, but I needed something in place to make progress

FabioLuporini replied:
should you not override `__eq__` inside `Derivative` instead?

JDBetteridge replied:
Yes, that is probably the right place, I think I fixed here as it was the source of the recursion exception. I think you probably still need something here to prevent infinite recursion

mloubout commented on devito/finite_differences/derivative.py:
```diff
@@ -262,7 +262,10 @@ def __call__(self, x0=None, fd_order=None, side=None, method=None, **kwargs):
         kwargs['preprocessed'] = True
-        return super()._rebuild(*args, **kwargs)
+        if args:
+            return super()._rebuild(diffify(args[0]), *args[1:], **kwargs)
```
This seems quite dangerous. Also `diffify` is usually not reconstructing things like `space_order` and such attributes properly so this will still loose a lot of information. 
You also probably better off doing that line 98

```
if not isinstance(expr, Differentiable):
    try:
        expr = diffify(expr)
    except not sure what:
        raise ValueError("`expr` must be a Differentiable object")
```

JDBetteridge replied:
I previously had that and I can't remember why I changed it, LMGBTY

JDBetteridge replied:
Moved, I potentially fixed the other issue elsewhere

mloubout commented:
Do you have some MFE (minimal failing example) of where this is popping up?

EdCaunt commented on devito/finite_differences/differentiable.py:
```diff
@@ -113,6 +113,73 @@ def is_Staggered(self):
+        A near copy of sympy.core.expr.Expr.as_independent
+        with a bug fixed
+        """
+        from sympy import Symbol
```
Can these imports be moved to the top of the file?

EdCaunt commented on devito/finite_differences/differentiable.py:
```diff
@@ -113,6 +113,73 @@ def is_Staggered(self):
+        else:
+            want = Mul
+
+        # sift out deps into symbolic and other and ignore
```
Nitpick: Start comment with capital level

EdCaunt commented on devito/finite_differences/differentiable.py:
```diff
@@ -113,6 +113,73 @@ def is_Staggered(self):
+                other.append(d)
+
+        def has(e):
+            """return the standard has() if there are no literal symbols, else
```
Nitpick: Tweak docstring for consistency

EdCaunt commented on devito/finite_differences/differentiable.py:
```diff
@@ -113,6 +113,73 @@ def is_Staggered(self):
+                return has_other
+            return has_other or e.has(*(e.free_symbols & sym))
+
+        if (want is not func or
```
Nitpick: Weird indent multiline

mloubout commented on devito/finite_differences/derivative.py:
```diff
@@ -258,10 +267,9 @@ def __call__(self, x0=None, fd_order=None, side=None, method=None, **kwargs):
             rkw['fd_order'] = self._filter_dims(_fd_order, as_tuple=True)
 
-        return self._rebuild(**rkw)
+        return self._rebuild(*self.args, **rkw)
```
`*self.args` should not be needed; that's what `rargs` is for at the class definition. `_rebuild` should only ever need the new arg/kwargs and inherit everything else from `self` automatically.

JDBetteridge commented:
Here is a wonderful list of failing examples that this PR now addresses:
```python
import devito as dv
import sympy as sym

grid = dv.Grid(shape=(11,), extent=(1,))
x = grid.dimensions[0]

u = dv.Function(name='u', grid=grid, space_order=4)

a = u.dx
b = a.subs({u: -5*u.dx + 4*u + 3})

# Substitution works!
print(b)

# Unintuitive results:

# .as_independent method on devito.finite_differences.differentiable.Mul doesn't work
print((-a).as_independent(x))
# Result:
## (1, -Derivative(u(x), x))

# devito.Derivatives aren't reconstructable from func and args (as per sympy docs)
du = u.dx
print(du.func(*du.args))
print(du.func(*du.args).args)
# Result
## Derivative(u(x), (x, 1))
## (u(x), ((x, 1), 1))

# default simplification causes different results:
du11 = dv.Derivative(u, x, x)
print(du11)
print(du11.deriv_order)
# Result
## Derivative(u(x), (x, 2))
## (1, 1)
du2 = dv.Derivative(u, (x, 2))
print(du2)
print(du2.deriv_order)
# Result
## Derivative(u(x), (x, 2))
## (2,)

# Whut!?
print(dv.Derivative(u, x, deriv_order=(2,4)))
# Result
## Derivative(u(x), (x, 2))

# Also pretty wild
print(dv.Derivative(u, (x, 0)))
# Result
## Derivative(u(x), x)

# Really? Double check this...
print(dv.Derivative(-1))
# Result
## Traceback (most recent call last):
##   File "<stdin>", line 1, in <module>
##   File "/media/devito/devito_py312/src/devito/devito/finite_differences/derivative.py", line 98, in __new__
##     raise ValueError("`expr` must be a Differentiable object")
## ValueError: `expr` must be a Differentiable object

# Maybe it's sufficient to handle the case:
print(dv.Derivative(sym.sympify(-1)))

# Cannot expand
print(b.expand())
# Result
## ...
## RecursionError: maximum recursion depth exceeded
```

EdCaunt commented on tests/test_derivatives.py:
```diff
@@ -1023,3 +1023,124 @@ def bypass_uneval(expr):
+        with pytest.raises(TypeError):
+            _ = Derivative(u, (x, a))
+
+    def test_expand_mul(self):
```
Do these work when using the derivative shortcuts?

mloubout commented on devito/finite_differences/derivative.py:
```diff
@@ -497,3 +505,57 @@ def _eval_fd(self, expr, **kwargs):
+        `Derivative(f(x) + g(x), x)
+            --> Derivative(f(x), x) + Derivative(g(x), x)`
+        '''
+        expr = self.args[0]
```
self.expr

mloubout commented on devito/finite_differences/derivative.py:
```diff
@@ -497,3 +505,57 @@ def _eval_fd(self, expr, **kwargs):
+        expr = self.args[0]
+        if isinstance(expr, sympy.Add):
+            ind, dep = expr.as_independent(*self.dims, as_Add=True)
+            if isinstance(dep, sympy.Add):
```
is_Add

mloubout commented on devito/finite_differences/differentiable.py:
```diff
@@ -113,6 +113,73 @@ def is_Staggered(self):
     def is_TimeDependent(self):
         return any(i.is_Time for i in self.dimensions)
 
+    def as_independent(self, *deps, **hint):
```
This seems quite brutal that we need that almost full copy paste here. Do we know which lines actually cause issue?

JDBetteridge replied:
I think this is resolved, open again if you aren't convinced

mloubout reviewed (commented):
Mostly nitpicky comments, looks like a nice cleanup

EdCaunt commented on devito/finite_differences/derivative.py:
```diff
@@ -92,19 +93,95 @@ class Derivative(sympy.Derivative, Differentiable, Pickable):
                    'x0', 'method', 'weights')
 
     def __new__(cls, expr, *dims, **kwargs):
+        # TODO: Delete this
```
Needs removing before merge

EdCaunt commented on devito/finite_differences/derivative.py:
```diff
@@ -92,19 +93,95 @@ class Derivative(sympy.Derivative, Differentiable, Pickable):
+            raise ValueError(f'Expected `(dim, deriv_order)`, got {dims[0]}')
+        elif len(dims) == 2 and not isinstance(dims[1], Iterable) and is_integer(dims[1]):
+            # special case of single dimension and order
+            dims = (dims, )
```
Just use `as_tuple`

JDBetteridge replied:
That doesn't work in this case, `dim` is already a `tuple`, but I want a nested tuple

EdCaunt commented on devito/finite_differences/derivative.py:
```diff
@@ -497,3 +505,57 @@ def _eval_fd(self, expr, **kwargs):
 
         return res
+
+    def _eval_expand_nest(self, **hints):
```
Would some kind of recursing `singledispatch` here tidy things up?

EdCaunt commented on devito/finite_differences/differentiable.py:
```diff
@@ -113,6 +113,73 @@ def is_Staggered(self):
+
+        if func is Add:  # all terms were treated as commutative
+            return (Add(*indep), _unevaluated_Add(*depend))
+        else:  # handle noncommutative by stopping at first dependent term
```
Nitpick: This file has a lot of comments which aren't in our nominal style

JDBetteridge commented on devito/finite_differences/derivative.py:
```diff
@@ -92,19 +93,95 @@ class Derivative(sympy.Derivative, Differentiable, Pickable):
+        # Default finite difference orders depending on input dimension (.dt or .dx)
+        # It's possible that the expr is a `sympy.Number` at this point, which
+        # has derivative 0, unless we're taking a 0th derivative.
+        if isinstance(expr, sympy.Number):
```
This seems like an optimisation, but actually is not the right thing to do here. Consider the case where you want to make a substitution by using a temporary. Now you cannot construct the `Derivative` as it will be prematurely simplified to zero, unless you explicitly construct the correct type of temporary.

I think in the case of just a number this is a safe thing to do.

JDBetteridge commented on devito/finite_differences/derivative.py:
```diff
@@ -497,3 +505,57 @@ def _eval_fd(self, expr, **kwargs):
 
         return res
+
+    def _eval_expand_nest(self, **hints):
```
No, we shouldn't be recursing here, see:
https://docs.sympy.org/latest/guides/custom-functions.html#expand
for reference. Sympy takes care of this for us.

EdCaunt commented on devito/finite_differences/derivative.py:
```diff
@@ -149,33 +150,58 @@ def __new__(cls, expr, *dims, **kwargs):
         else:
-            default_fdo = tuple([
+            # Default finite difference orders depending on input dimension (.dt or .dx)
+            fd_order = tuple([
```
Nitpick: Unnecessary square brackets

EdCaunt commented on devito/finite_differences/derivative.py:
```diff
@@ -149,33 +150,58 @@ def __new__(cls, expr, *dims, **kwargs):
+        if fd_order is not None:
+            # If `fd_order` is specified collect these together
+            fcounter = defaultdict(int)
+            for d, o in zip(dims, as_tuple(fd_order)):
```
Nitpick: Some extra comments on this big stack of loops and conditionals would help, since my eyes are glazing over looking at it

EdCaunt commented on devito/finite_differences/derivative.py:
```diff
@@ -520,21 +546,56 @@ def _eval_expand_nest(self, **hints):
             # EG: Don't actually expand if derivatives are incompatible
-            new_kwargs = {'deriv_order': tuple(chain(self.deriv_order, expr.deriv_order))}
-            return self.func(new_expr, *new_dims, **new_kwargs)
+            new_deriv_order = tuple(chain(self.deriv_order, self.expr.deriv_order))
```
`chain(self.deriv_order, self.expr.deriv_order)` is used above, so perhaps define it as a variable to improve readability?

EdCaunt commented on devito/finite_differences/derivative.py:
```diff
@@ -555,12 +615,23 @@ def _eval_expand_add(self, **hints):
+        resultant expression for higher derivatives and mixed derivatives is much
+        more difficult to implement.
+        '''
+        # if self.expr.is_Mul:
```
Leftover

EdCaunt commented on devito/finite_differences/derivative.py:
```diff
@@ -555,12 +615,23 @@ def _eval_expand_add(self, **hints):
+        This is only implemented for first derivatives with an arbitrary number
+        of multiplicands and second derivatives with two multiplicands. The
+        resultant expression for higher derivatives and mixed derivatives is much
+        more difficult to implement.
```
Guessing it needs some variant on https://en.wikipedia.org/wiki/Fa%C3%A0_di_Bruno%27s_formula ?

EdCaunt commented on devito/finite_differences/differentiable.py:
```diff
@@ -503,6 +440,61 @@ def has_free(self, *patterns):
+
+
+# Monkeypatch the method
+if Version(sympy.__version__) < Version('1.15.0.dev0'):
```
Should this live in `devito/mpatches`?

mloubout reviewed (commented):
few more comments

mloubout commented on devito/finite_differences/derivative.py:
```diff
@@ -497,3 +580,141 @@ def _eval_fd(self, expr, **kwargs):
+        more difficult to implement.
+        """
+        if self.expr.is_Mul and len(self.dims) == 1:
+            if self.deriv_order == (1,):
```
You can do `args=self.expr.args` before those to make it less verbose and avoid the `self.expr.args` everywhere

mloubout commented on devito/finite_differences/derivative.py:
```diff
@@ -497,3 +580,141 @@ def _eval_fd(self, expr, **kwargs):
+                ])
+            elif self.deriv_order == (2,) and len(self.expr.args) == 2:
+                return Add(
+                    Mul(self.func(self.expr.args[0]), self.expr.args[1]),
```
nitpicking: don't really need `Mul/Add` here, just `*` is fine Differentiable will take care of it

JDBetteridge replied:
Sure, the one above I'm leaving as is though

mloubout commented on devito/finite_differences/derivative.py:
```diff
@@ -497,3 +580,141 @@ def _eval_fd(self, expr, **kwargs):
+        # actually expanded if derivatives are incompatible.
+        # The nested derivative is evaluated by:
+        # 1. Chaining together the variables with which to differentiate wrt
+        new_expr = self.expr.args[0]
```
self.expr.expr

mloubout commented on devito/finite_differences/derivative.py:
```diff
@@ -497,3 +580,141 @@ def _eval_fd(self, expr, **kwargs):
+            for d, ii in zip(
+                chain(self.dims, self.expr.dims),
+                chain(self.deriv_order, self.expr.deriv_order),
+                strict=True
```
?

mloubout replied:
the `strict` yes

JDBetteridge replied:
This ensures that the two iterables are of the same length

https://docs.python.org/3/library/functions.html#zip

New in 3.10

JDBetteridge replied:
I think this is resolved, open again if you aren't convinced

mloubout commented on devito/finite_differences/derivative.py:
```diff
@@ -497,3 +580,141 @@ def _eval_fd(self, expr, **kwargs):
+                order = self.expr.space_order
+            # The `fd_order` may need to be reduced to construct the nested derivative
+            # in this case we only emit a warning
+            if fo > order:
```
I'm not sure how we can end up here. If the nested derivative is built and considered acceptable I'm not sure we should throw an error here.

JDBetteridge replied:
I think this is resolved, open again if you aren't convinced

mloubout commented on devito/finite_differences/derivative.py:
```diff
@@ -497,3 +580,141 @@ def _eval_fd(self, expr, **kwargs):
+                        f'than the {dim_name}_order={order}'
+                    )
+                else:
+                    warn(
```
why would the order be increased here? If the user specified a smaller  fd_order it should be kept.

JDBetteridge replied:
If the user is taking the `Derivative(..., fd_order=2)` of a `Derivative(..., fd_order=2)` what should the result be? I think you want an `fd_order=4` derivative, but maybe not?

JDBetteridge replied:
I think this is resolved, open again if you aren't convinced

mloubout reviewed (commented):
Some minor/nitpicky comments but looking good

enwask commented on devito/mpatches/asindependent.py:
```diff
@@ -0,0 +1,73 @@
+"""Monkeypatch for as_independent required for Devito Derviative. """
```
Is that the British spelling?

JDBetteridge replied:
No, I just can't spell

enwask reviewed (commented):
Small typo

EdCaunt commented on devito/mpatches/asindependent.py:
`as_independent.py`?

JDBetteridge commented on devito/finite_differences/derivative.py:
```diff
@@ -125,73 +155,122 @@ def __new__(cls, expr, *dims, **kwargs):
+        dcounter = defaultdict(int)
+        for d, o in zip(dims, deriv_order, strict=True):
+            if isinstance(d, Iterable):
+                if not is_integer(d[1]) or d[1] < 0:
```
Mainly to prevent symbolic derivative orders being passed in (unsupported)

JDBetteridge commented on devito/finite_differences/derivative.py:
```diff
@@ -125,73 +155,122 @@ def __new__(cls, expr, *dims, **kwargs):
-                          for s in odims]
-        return new_dims, orders, fd_orders, variable_count
+                    fcounter[d] += o
+            # Second validate that the `fd_order` is supported by the space or
```
I think this is better than the previous behaviour which would just cause the derivative to evaluate to zero. But this is up for discussion...

JDBetteridge replied:
I think this is resolved, open again if you aren't convinced

JDBetteridge commented:
Results of profiling:

See all resources in the compressed folder

```
$ ./profiling_demo.sh | tee output.txt
Switched to branch 'worktree_py312'
Your branch is up to date with 'main'.
Operator `Kernel` generated in 13.10 s
  * lowering.Clusters: 8.07 s (61.7 %)
     * specializing.Clusters: 6.26 s (47.9 %)
        * cire: 3.79 s (29.0 %)
  * lowering.Expressions: 4.70 s (35.9 %)
Flops reduction after symbolic optimization: [10128 --> 215]
Operator `Kernel` generated in 12.88 s
  * lowering.Clusters: 8.11 s (63.0 %)
     * specializing.Clusters: 6.19 s (48.1 %)
        * cire: 3.95 s (30.7 %)
  * lowering.Expressions: 4.46 s (34.7 %)
Flops reduction after symbolic optimization: [10128 --> 215]
Operator `Kernel` generated in 2.29 s
  * lowering.Clusters: 1.44 s (63.0 %)
     * specializing.Clusters: 0.95 s (41.5 %)
        * cire: 0.69 s (30.2 %)
  * lowering.Expressions: 0.59 s (25.8 %)
Flops reduction after symbolic optimization: [7302 --> 291]

real	0m29.538s
user	0m31.264s
sys	0m0.277s
Operator `Kernel` generated in 30.31 s
  * lowering.Clusters: 19.87 s (65.6 %)
     * specializing.Clusters: 15.42 s (50.9 %)
        * cire: 9.47 s (31.3 %)
  * lowering.Expressions: 9.71 s (32.1 %)
Flops reduction after symbolic optimization: [10128 --> 215]
Operator `Kernel` generated in 29.79 s
  * lowering.Clusters: 19.68 s (66.1 %)
     * specializing.Clusters: 15.13 s (50.8 %)
        * cire: 9.55 s (32.1 %)
  * lowering.Expressions: 9.39 s (31.6 %)
Flops reduction after symbolic optimization: [10128 --> 215]
Operator `Kernel` generated in 5.39 s
  * lowering.Clusters: 3.52 s (65.4 %)
     * specializing.Clusters: 2.40 s (44.6 %)
        * cire: 1.78 s (33.1 %)
  * lowering.Expressions: 1.21 s (22.5 %)
Flops reduction after symbolic optimization: [7302 --> 291]

real	1m12.504s
user	1m13.780s
sys	0m0.482s
py-spy> Sampling process 500 times a second. Press Control-C to exit.

Operator `Kernel` generated in 25.37 s
  * lowering.Clusters: 15.15 s (59.8 %)
     * specializing.Clusters: 11.92 s (47.1 %)
        * cire: 7.05 s (27.8 %)
  * lowering.Expressions: 9.55 s (37.7 %)
Flops reduction after symbolic optimization: [10128 --> 215]
Operator `Kernel` generated in 25.28 s
  * lowering.Clusters: 15.65 s (62.0 %)
     * specializing.Clusters: 12.46 s (49.3 %)
        * cire: 7.34 s (29.1 %)
  * lowering.Expressions: 9.00 s (35.7 %)
Flops reduction after symbolic optimization: [10128 --> 215]
Operator `Kernel` generated in 4.38 s
  * lowering.Clusters: 2.55 s (58.3 %)
     * specializing.Clusters: 1.51 s (34.6 %)
        * cire: 1.04 s (23.8 %)
  * lowering.Expressions: 1.25 s (28.6 %)
Flops reduction after symbolic optimization: [7302 --> 291]

py-spy> Stopped sampling because process exited
py-spy> Wrote flamegraph data to 'main_pys.svg'. Samples: 28478 Errors: 35
Error: No child process (os error 10)

real	0m57.425s
user	0m40.719s
sys	0m28.169s
Switched to branch 'JDBetteridge/expand_derivative'
Your branch and 'origin/JDBetteridge/expand_derivative' have diverged,
and have 143 and 31 different commits each, respectively.
  (use "git pull" if you want to integrate the remote branch with yours)
Operator `Kernel` generated in 13.09 s
  * lowering.Clusters: 8.11 s (62.0 %)
     * specializing.Clusters: 6.29 s (48.1 %)
        * cire: 3.91 s (29.9 %)
  * lowering.Expressions: 4.66 s (35.7 %)
Flops reduction after symbolic optimization: [10128 --> 215]
Operator `Kernel` generated in 12.96 s
  * lowering.Clusters: 8.23 s (63.6 %)
     * specializing.Clusters: 6.30 s (48.7 %)
        * cire: 4.01 s (31.0 %)
  * lowering.Expressions: 4.42 s (34.2 %)
Flops reduction after symbolic optimization: [10128 --> 215]
Operator `Kernel` generated in 2.31 s
  * lowering.Clusters: 1.44 s (62.5 %)
     * specializing.Clusters: 0.95 s (41.3 %)
        * cire: 0.68 s (29.6 %)
  * lowering.Expressions: 0.60 s (26.1 %)
Flops reduction after symbolic optimization: [7302 --> 291]

real	0m29.629s
user	0m31.293s
sys	0m0.267s
Operator `Kernel` generated in 29.64 s
  * lowering.Clusters: 19.37 s (65.4 %)
     * specializing.Clusters: 14.98 s (50.6 %)
        * cire: 9.15 s (30.9 %)
  * lowering.Expressions: 9.55 s (32.3 %)
Flops reduction after symbolic optimization: [10128 --> 215]
Operator `Kernel` generated in 29.27 s
  * lowering.Clusters: 19.12 s (65.4 %)
     * specializing.Clusters: 14.70 s (50.3 %)
        * cire: 9.26 s (31.7 %)
  * lowering.Expressions: 9.47 s (32.4 %)
Flops reduction after symbolic optimization: [10128 --> 215]
Operator `Kernel` generated in 5.19 s
  * lowering.Clusters: 3.34 s (64.5 %)
     * specializing.Clusters: 2.29 s (44.2 %)
        * cire: 1.70 s (32.8 %)
  * lowering.Expressions: 1.22 s (23.6 %)
Flops reduction after symbolic optimization: [7302 --> 291]

real	1m10.286s
user	1m11.559s
sys	0m0.511s
py-spy> Sampling process 500 times a second. Press Control-C to exit.

Operator `Kernel` generated in 25.57 s
  * lowering.Clusters: 14.91 s (58.4 %)
     * specializing.Clusters: 11.76 s (46.1 %)
        * cire: 7.12 s (27.9 %)
  * lowering.Expressions: 9.82 s (38.5 %)
Flops reduction after symbolic optimization: [10128 --> 215]
Operator `Kernel` generated in 24.50 s
  * lowering.Clusters: 15.30 s (62.5 %)
     * specializing.Clusters: 12.01 s (49.1 %)
        * cire: 7.02 s (28.7 %)
  * lowering.Expressions: 8.66 s (35.4 %)
Flops reduction after symbolic optimization: [10128 --> 215]
Operator `Kernel` generated in 4.20 s
  * lowering.Clusters: 2.44 s (58.2 %)
     * specializing.Clusters: 1.41 s (33.6 %)
        * cire: 0.92 s (22.0 %)
  * lowering.Expressions: 1.17 s (27.9 %)
Flops reduction after symbolic optimization: [7302 --> 291]

py-spy> Stopped sampling because process exited
py-spy> Wrote flamegraph data to 'expand_pys.svg'. Samples: 28088 Errors: 28
Error: No child process (os error 10)

real	0m56.759s
user	0m39.889s
sys	0m27.775s
```
![main_pys](https://github.com/user-attachments/assets/22ef25f6-d9ec-4c43-a626-3327b41f529c)
![expand_pys](https://github.com/user-attachments/assets/f6446a75-fccc-4ff1-816b-053972ea3dde)

[profiling.zip](https://github.com/user-attachments/files/21510690/profiling.zip)

mloubout commented on devito/finite_differences/derivative.py:
```diff
@@ -261,7 +339,6 @@ def __call__(self, x0=None, fd_order=None, side=None, method=None, **kwargs):
         return self._rebuild(**rkw)
 
     def _rebuild(self, *args, **kwargs):
-        kwargs['preprocessed'] = True
```
Just `func = Pickable._rebuild` and you can remove `_rebuild`

JDBetteridge commented on devito/finite_differences/derivative.py:
```diff
@@ -92,19 +94,47 @@ class Derivative(sympy.Derivative, Differentiable, Pickable):
-            raise ValueError("Cannot nest sympy.Derivative with devito.Derivative")
-        if not isinstance(expr, Differentiable):
-            raise ValueError("`expr` must be a Differentiable object")
+        # TODO: Delete this
+        if kwargs.get('preprocessed', False):
+            from warnings import warn as pywarn
+            pywarn('I removed the `preprocessed` kwarg')
+
```
```suggestion
```
